### PR TITLE
CLIC 2020 Image Compression dataset

### PR DESCRIPTION
**Pull request by 0x00b1**
CLIC 2020 Image Compression dataset

### CODE REVIEW

File: neuralcompression/data/clic2020.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import torchvision.datasets.folder
+import torchvision.datasets.utils
+
+
+class CLIC2020(torch.utils.data.Dataset):
+    """`Challenge on Learned Image Compression (CLIC) 2020
+    <http://compression.cc/tasks/>`_ Dataset.
+
+    Args:
+        root (str): Root directory where images are downloaded to.
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:17]: No need for type annotations in docstrings (see sphinx PR).

File: neuralcompression/data/clic2020.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+                <root>
+                    └── clic2020
+                        ├── test
+                        │   ├── *.png
+                        ├── train
+                        │   ├── *.png
+                        └── val
+                            └── *.png
+        split (str): The dataset split to use. One of
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:30]: No need for `str`.

File: neuralcompression/data/clic2020.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            {``train``, ``val``, ``test``}.
+            Defaults to ``train``.
+        download (bool, optional): If true, downloads the dataset from the
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:33]: No need for `bool, optional`.

File: neuralcompression/data/clic2020.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            internet and puts it in root directory. If dataset is already
+            downloaded, it is not downloaded again.
+        transform (callable, optional): A function/transform that takes in a
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:36]: No need for `callable, optional`.

File: neuralcompression/data/clic2020.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            "url": "https://storage.googleapis.com/clic2021_public/professional_test_2021.zip",
+        },
+    }
+
+    def __init__(
+        self,
+        root: typing.Union[str, pathlib.Path],
+        split: str = "train",
+        download: bool = False,
+        transform: typing.Optional[typing.Callable] = None,
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:64]: Could annotate further specify this as a `typing.Callable[PIL.Image, torch.Tensor]`.

File: neuralcompression/data/clic2020.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            self.download()
+
+        self._paths = None
+
+    @property
+    def paths(self):
+        if self._paths:
+            return self._paths
+        else:
+            return [*self.root.joinpath(self.split).glob("*.png")]
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:84]: Why not define this at init time? Is this just a difference between the images available when `__init__` is called and the first `__getitem__`? I don't see the added benefit.

File: tests/test_clic2020.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+from neuralcompression.data import CLIC2020
+
+
+@pytest.fixture(autouse=True)
+def remove_data_directory():
+    yield
+
+    if pathlib.Path("data").exists():
+        shutil.rmtree("data")
```
Comment (mmuckley) [inline tests/test_clic2020.py:14]: No need for this - `pytest` has a temporary data directory. See https://github.com/facebookresearch/NeuralCompression/blob/e27c96287cdf06013f38df8913e08c7b8c09e42e/tests/test_kodak.py#L27-L48

File: tests/test_clic2020.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        with pytest.raises(IndexError):
+            assert test_data[0]
+
+    def test___len__(self, test_data):
+        assert len(test_data) == 0
+
+    def test_download(self, test_data):
+        test_data.download()
+
+        assert len(test_data) == 60
```
Comment (mmuckley) [inline tests/test_clic2020.py:33]: We shouldn't include downloads in CI - we can just trust that you ran it and it worked. See the reference linked above for an example of creating test data and monkeypatching.

**Review by mmuckley — CHANGES_REQUESTED** [review]
Requesting a few changes.

- No need for type annotations in docstrings.
- Confusing use of the `paths` property - I'm not sure of the benefit. Why not just define them in `__init__`?
- Please avoid downloads in CI. There's an example of data mocking in the linked test.

File: neuralcompression/data/clic2020.py (hunk 6 above)
Comment (0x00b1) [reply neuralcompression/data/clic2020.py]: It was old and should've been refactored out!

File: neuralcompression/data/clic2020.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
+import os
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

File: tests/test_clic2020.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import pathlib
```
Comment (mmuckley) [inline tests/test_clic2020.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

**Review by mmuckley — CHANGES_REQUESTED** [review]

File: neuralcompression/data/clic2020.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        },
+        "val": {
+            "filename": "val.zip",
+            "md5": "7111ee240435911db04dbc5f40d50272",
+            "url": "https://data.vision.ee.ethz.ch/cvl/clic/professional_valid_2020.zip",
+        },
+        "test": {
+            "filename": "test.zip",
+            "md5": "4476b708ea4c492505dd70061bebe202",
+            "url": "https://storage.googleapis.com/clic2021_public/professional_test_2021.zip",
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:62]: nit: break up this string?

File: neuralcompression/data/clic2020.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        kwargs = {
+            "download_root": str(self.root),
+            "extract_root": str(self.root),
+            **self.splits[self.split],
+        }
+
+        if self.split == "test":
+            kwargs["extract_root"] = self.root.joinpath("test")
+
+        torchvision.datasets.utils.download_and_extract_archive(**kwargs)
```
Comment (mmuckley) [inline neuralcompression/data/clic2020.py:111]: nit: use standard keyword argument syntax instead of `kwargs`?

File: tests/test_clic2020.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+from utils import create_random_image
+
+
+@pytest.fixture
+def data(tmp_path):
+    directory = tmp_path.joinpath("clic2020").joinpath("test")
+
+    directory.mkdir(parents=True)
+
+    for index in range(3):
```
Comment (mmuckley) [inline tests/test_clic2020.py:21]: Should 3 be a parameter? Tests for other splits?

**Review by mmuckley — APPROVED** [review]
Made a few comments but nothing blocking. Nice work!

**Comment by 0x00b1** [comment]
@mmuckley I went ahead and resolved your comments. I also changed the import style.

**Pull request merged**